elk: 7.5.1 -> 7.11.1

thefloweringash opened the pull request:
###### Motivation for this change

Package update and support for aarch64-linux, but unfortunately not a simple update. This requires touching a lot of related packages, most of which I'm not very familiar with. Hopefully I can at least I can get the ball rolling on the update.

Notably:

- Drops elasticsearch7-oss and kibana7-oss, since they're no longer distributed by upstream as of 7.11. We might be able to provide these under the SSPL, but they wouldn't fit into the current pull request.
- ~Breaks `search-guard` and `elasticsearch-lemmagen` plugins, which don't seem support elasticsearch 7.11.x yet.~ Plugins updated!

cc maintainers: @apeschar, @basvandijk, @wjlroe, @offlinehacker

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS (aarch64-linux and x86_64-linux)
   - [x] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

thefloweringash committed:
elk7Version: 7.5.1 -> 7.11.1

thefloweringash committed:
elasticsearch/plugins: 7.5.1 -> 7.11.1

thefloweringash committed:
elasticsearch7: 7.5.1 -> 7.11.1; drop enableUnfree

thefloweringash committed:
logstash7: 7.5.1 -> 7.11.1

thefloweringash committed:
kibana7: 7.5.1 -> 7.11.1; drop enableUnfree

thefloweringash committed:
beats7: 7.5.1 -> 7.11.1

thefloweringash committed:
nixos/tests/elk: propagate pkgs and system to test runner

thefloweringash committed:
nixos/tests/elk: remove OSS ELK-7 test

SuperSandro2000 commented on pkgs/servers/search/elasticsearch/7.x.nix:
```diff
@@ -46,38 +41,43 @@ stdenv.mkDerivation (rec {
-  passthru = { inherit enableUnfree; };
+  postFixup = lib.optionalString stdenv.hostPlatform.isLinux ''
+    for exe in $(find $out/modules/x-pack-ml/platform/${plat}-${arch}/bin -executable -type f); do
+      echo "patching $exe..."
```
Doesn't patchelf echo what it is doing?

thefloweringash replied:
I don't see what this `postFixup` hook is doing that `autoPatchelfHook` doesn't, so I'm now removing it.

thefloweringash committed:
elasticsearch7: remove manual elf patching

The options as specified were not a coherent set. There were three
things to consider: autoPatchelfHook, the regular rpath
fixup (controlled by dontPatchELF) and the elf interpreter rewrite in
the postFixup hook.

The autoPatchelfHook will set the interpreter, so the explicit
invocation of patchelf to do so in postFixup should not be required.

The autoPatchelfHook will rewrite rpaths entirely, so disabling the
rpath minimizing via dontPatchELF should have no effect.

thefloweringash commented:
I've marked this as ready for review, but would appreciate additional testing and the maintainer's input here.

SuperSandro2000 commented on pkgs/misc/logging/beats/7.x.nix:
```diff
@@ -1,17 +1,17 @@
 
-let beat = package : extraArgs : buildGoPackage (rec {
+let beat = package : extraArgs : buildGoModule (rec {
       name = "${package}-${version}";
```
That could be converted to pname + version.

SuperSandro2000 commented on pkgs/servers/search/elasticsearch/7.x.nix:
```diff
@@ -15,19 +14,15 @@ let
+stdenv.mkDerivation rec {
   version = elk7Version;
-  name = "elasticsearch-${optionalString (!enableUnfree) "oss-"}${version}";
+  name = "elasticsearch-${version}";
```
```suggestion
  pname = "elasticsearch";
```

thefloweringash commented:
Pushed an update to use `pname` everywhere I saw `name`.

I've been using this expression to keep a track of the packages changed and sources this update requires: [gist](https://gist.github.com/thefloweringash/8d359cadb8f3c8c32f3d96c896ac9b3a).

zbioe commented:
Thank's for this PR. I really appreciate it

thefloweringash committed:
elasticsearch/plugins: build with "modules" from elasticsearch

Fixes dependency error while building search-guard:

    Exception in thread "main" java.lang.IllegalArgumentException: Missing plugin [lang-painless], dependency of [search-guard-7]

thefloweringash commented:
I've updated this to fix merge conflicts only.

Since I opened this the versions have changed. The current version elasticsearch is 7.13, but some plugins are limited (searchguard goes up to 7.12.1, lemmagen goes up to 7.12). I don't know what's best here: to hold back elasticsearch, or to keep the maximum plugin compatibility.

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -4367,11 +4367,6 @@ in
     util-linux = util-linuxMinimal;
     jre_headless = jre8_headless; # TODO: remove override https://github.com/NixOS/nixpkgs/pull/89731
   };
-  elasticsearch7-oss = callPackage ../servers/search/elasticsearch/7.x.nix {
```
We need aliases entries for the removals.

doronbehar commented:
@thefloweringash it seems you have addressed all review comments, there's a small merge conflict now. 

As for:

> but some plugins are limited (searchguard goes up to 7.12.1, lemmagen goes up to 7.12). I don't know what's best here: to hold back elasticsearch, or to keep the maximum plugin compatibility.

It seems not too many plugins are incompatible. I think you should update elasticsearch to the latest, and mark these plugins as broken for elasticsearch 7.3, as done here:

https://github.com/NixOS/nixpkgs/blob/3d72b0b6b0ff8a4cd46ebad0382d8d06d7255e95/pkgs/servers/search/elasticsearch/plugins.nix#L38-L41

happysalada commented:
I'm only seeing this PR now.
I've made the upgrade to elk 7.10.2 which is the last version including oss.
Most of the fixes in this PR are integrated, except the hashes for version 7.11.1 of course.
There is one more weird fix needed to get elasticsearch working properly, the jvm gc.log option needs to be a writeable path.
I've got the fix in this PR https://github.com/NixOS/nixpkgs/pull/136255
There are 2 things that I'm not sure about. If anybody is interested, it could use some feedback.

roberth committed:
Revert "nixos/tests/elk: remove OSS ELK-7 test"

This reverts commit 35db6bc5543d352ea7e54375ba12acd70033a7f9.

roberth committed:
nixos/tests/elk: Restructure attribute layout to avoid ELK-7 (non-OSS) on Hydra

... or any other tool that respects the recurseForDerivations attribute.

roberth commented:
I've rebased and successfully tested `nixosTests.elk.unfree`.

roberth commented:
I've rebased and successfully tested `nixosTests.elk.unfree`.

roberth committed:
elasticsearch/plugin: Fix rebase error
